File: cf-support/cmd/pipelines.py
from logic import codefresh, core
from logic.k8s import select_namespace
from utils import resource_list, files
import time

# import argparse


def setup_parser(parser):
    pass


def execute(args):
    runtime_type = "pipelines"
    dir_path = f"cf-support-{runtime_type}-{int(time.time())}"
    cf_config = codefresh.get_creds()
    runtimes = codefresh.get_saas_runtimes(cf_config)

    if len(runtimes) != 0:
        for index, runtime in enumerate(runtimes, start=1):
            print(f"{index}. {runtime['metadata']['name']}")

        while True:
            try:
                selection = int(
                    input("\nPlease select the runtime to gather data from (Number): ")
                )
                if 1 <= selection <= len(runtimes):
                    break
                else:
                    print("Invalid selection. Please enter a valid number.")
            except ValueError:
                print("Invalid input. Please enter a valid number.")

        re_spec = runtimes[selection - 1]
        pipelines_namespace = re_spec["runtimeScheduler"]["cluster"]["namespace"]
    else:
        print("No runtimes found in the Codefresh account.")
        pipelines_namespace = select_namespace(runtime_type)

    print(f"Gathering data in {pipelines_namespace} namespace")
    k8s_resources = {**resource_list.k8s_general, **resource_list.k8s_classic}
    core.gather_data(pipelines_namespace, k8s_resources, dir_path)

    if re_spec:
        files.save_file(files.to_yaml(re_spec), "pipelines-runtime-spec.yaml", dir_path)

    print("Data gathered successfully.")
    files.compress_dir(dir_path)


if __name__ == "__main__":
    # Example of how you might test the command directly
    # parser = argparse.ArgumentParser(description=__doc__)
    # setup_parser(parser)
    # args = parser.parse_args()
    # execute(args)
    execute(None)

File: src/pipelines.js
import { getResources, selectNamespace } from './logic/k8s.js';
import { preparePackage, processData, writeYaml } from './logic/core.js';
import { getAccountRuntimes, getCodefreshCredentials, getRuntimeSpec } from './logic/codefresh.js';

export async function pipelines(namespace, runtime) {
    const dirPath = `./cf-support-gitops-${Math.floor(Date.now() / 1000)}`;
    const cfCreds = getCodefreshCredentials();

    if (!namespace) {
        const selected = await selectNamespace();
        namespace = selected;
    }

    if (cfCreds) {
        if (!runtime) {
            const runtimes = await getAccountRuntimes(cfCreds);

            if (runtimes.length !== 0) {
                runtimes.forEach((re, index) => {
                    console.log(`${index + 1}. ${re.metadata.name}`);
                });
                let selection;
                do {
                    selection = Number(prompt('\nWhich Pipelines Runtime Are We Working With? (Number): '));
                    if (isNaN(selection) || selection < 1 || selection > runtimes.length) {
                        console.log(
                            'Invalid selection. Please enter a number corresponding to one of the listed runtimes.',
                        );
                    }
                } while (isNaN(selection) || selection < 1 || selection > runtimes.length);

                reSpec = runtimes[selection - 1];
                writeYaml(reSpec, 'Runtime_Spec', dirPath);
            }
        } else {
            reSpec = getRuntimeSpec(cfCreds, runtime);
            writeYaml(reSpec, 'Runtime_Spec', dirPath);
        }
    }

    console.log(`Gathering data in the '${namespace}' namespace for Codefresh OnPrem`);
    const k8sResources = getResources(namespace);
    await processData(dirPath, k8sResources);
    await preparePackage(dirPath);
}
